tests/Aspire.EndToEnd.Tests/IntegrationServicesFixture.cs

```diff
@@ -2,40 +2,92 @@
-/// 
+///
 /// Represents the the IntegrationServiceA project in the test application used to send HTTP requests
 /// to the project's endpoints.
 /// </summary>
 public sealed class IntegrationServicesFixture : IAsyncLifetime
 {
-    private Process? _appHostProcess;
-    private Dictionary<string, ProjectInfo>? _projects;
+#if TESTS_RUNNING_OUTSIDE_OF_REPO
+    public static bool TestsRunningOutsideOfRepo = true;
+#else
+    public static bool TestsRunningOutsideOfRepo;
+#endif
 
     public Dictionary<string, ProjectInfo> Projects => _projects!;
-
+    public BuildEnvironment BuildEnvironment { get; init; }
     public ProjectInfo IntegrationServiceA => Projects["integrationservicea"];
 
+    private Process? _appHostProcess;
+    private readonly TaskCompletionSource _appExited = new();
+    private Dictionary<string, ProjectInfo>? _projects;
+    private readonly IMessageSink _diagnosticMessageSink;
+    private readonly TestOutputWrapper _testOutput;
+
+    public IntegrationServicesFixture(IMessageSink diagnosticMessageSink)
+    {
+        _diagnosticMessageSink = diagnosticMessageSink;
+        _testOutput = new TestOutputWrapper(messageSink: _diagnosticMessageSink);
+        BuildEnvironment = new(TestsRunningOutsideOfRepo, (probePath, solutionRoot) =>
+        {
+            throw new InvalidProgramException(
+                    $"Running outside-of-repo: Could not find {probePath} computed from solutionRoot={solutionRoot}. " +
+                    $"Build all the packages with `./build -pack`. And install the sdk+workload 'dotnet build tests/Aspire.EndToEnd.Tests/Aspire.EndToEnd.csproj /t:InstallWorkloadUsingArtifacts /p:Configuration=<config>");
+        });
+        if (BuildEnvironment.HasSdkWithWorkload)
+        {
+            BuildEnvironment.EnvVars["TestsRunningOutsideOfRepo"] = "true";
+        }
+    }
+
     public async Task InitializeAsync()
     {
-        var appHostDirectory = Path.Combine(GetRepoRoot(), "tests", "testproject", "TestProject.AppHost");
+        var appHostDirectory = Path.Combine(BuildEnvironment.TestProjectPath, "TestProject.AppHost");
+        if (TestsRunningOutsideOfRepo)
+        {
+            _testOutput.WriteLine("");
+            _testOutput.WriteLine($"****************************************");
+            _testOutput.WriteLine($"   Running tests outside-of-repo");
+            _testOutput.WriteLine($"   TestProject: {appHostDirectory}");
+            _testOutput.WriteLine($"   Using dotnet: {BuildEnvironment.DotNet}");
+            _testOutput.WriteLine($"****************************************");
+            _testOutput.WriteLine("");
+        }
+
+        {
+            using var cmd = new DotNetCommand(BuildEnvironment, _testOutput, label: "build")
+                .WithWorkingDirectory(appHostDirectory);
+
+            (await cmd.ExecuteAsync(CancellationToken.None, $"build -bl:{Path.Combine(BuildEnvironment.LogRootPath, "testproject-build.binlog")} -v m"))
```

Comment:
Why `-v m`?

Reply:
Because most of the time the build is going to pass and doesn't give any useful info for issues in tests. If the build fails then those messages will show up, and a `testproject-build.binlog` will be available for debugging offline.